### PR TITLE
fix(qemu): pass SLIRP DNS address via kernel cmdline when QEMU_NET_CIDR is set

### PR DESCRIPTION
**Pull request by jmeridth**
## What

When QEMU_NET_CIDR overrides the SLIRP internal network, derive the correct DNS server address (network base + 3, per SLIRP convention) and pass it to the guest via the `dns=` kernel command line parameter.

## Why

Guest init scripts typically write `/etc/resolv.conf` using the `dns=` kernel param with a fallback to `10.0.2.3` (SLIRP's default). When QEMU_NET_CIDR moves SLIRP to a different subnet (e.g., `192.168.76.0/24`), `10.0.2.3` no longer exists on that network and all DNS resolution inside the QEMU VM fails. This breaks git clones, submodule fetches, and any other network operation requiring name resolution.

## Notes

- When QEMU_NET_CIDR is not set or empty, no `dns=` is added to the kernel cmdline. Zero behavior change for existing builds.
- The `dns=` parameter follows QEMU/Linux kernel cmdline conventions and is already consumed by common guest init scripts.
- `slirpDNSAddr` uses proper uint32 arithmetic to derive the address and validates the result is within the network.

## Testing

- [x] `go test ./pkg/container/ -v` -- all existing tests pass, no regressions
- [x] `go vet ./pkg/container/` -- clean
- [x] New `TestSlirpDNSAddr` covers: /24, /16, /8 networks, default SLIRP network, invalid input, empty input

### CODE REVIEW

File: pkg/container/qemu_runner.go [hunk 1]
```diff
@@ -2589,6 +2602,40 @@ func parseAndValidateNetCIDR(input string) (string, error) {
...
+
+	// Use the network address (mask away host bits) and add 3.
+	base := ipnet.IP.To4()
+	if base == nil {
+		return "", fmt.Errorf("network address is not IPv4")
+	}
+
+	// Convert to uint32 for safe arithmetic, then add 3.
+	n := uint32(base[0])<<24 | uint32(base[1])<<16 | uint32(base[2])<<8 | uint32(base[3])
+	n += 3
```
Comment (egibs) [inline pkg/container/qemu_runner.go]: This `base + 3` derivation only matches QEMU for /23–/30. For /8–/22 QEMU assigns `net | (0x0203 & ~mask)` (the third octet picks up `+2`), so the value written to the guest's `resolv.conf` won't match the resolver SLIRP is actually serving.

File: pkg/container/qemu_runner.go [hunk 2]
```diff
@@ -806,12 +806,18 @@ func createMicroVM(ctx context.Context, cfg *Config) error {
...
 			return fmt.Errorf("invalid QEMU_NET_CIDR value %q: %w", netCIDR, err)
 		}
-		log.Infof("qemu: QEMU_NET_CIDR set to %s, overriding SLIRP default network", cidr)
+		dnsIP, err := slirpDNSAddr(cidr)
+		if err != nil {
+			return fmt.Errorf("deriving SLIRP DNS address from QEMU_NET_CIDR %q: %w", cidr, err)
+		}
+		slirpDNS = dnsIP
+		log.Infof("qemu: QEMU_NET_CIDR set to %s, overriding SLIRP default network (dns=%s)", cidr, slirpDNS)
 		netdevArgs += ",net=" + cidr
```
Comment (egibs) [inline pkg/container/qemu_runner.go]: Consider appending `",dns=" + slirpDNS` here as well. Right now melange sets only `net=` and then independently guesses the DNS address QEMU will choose so setting `dns=` explicitly makes melange authoritative and guarantees that `netdev` and `cmdline` agree for every prefix length.

File: pkg/container/qemu_runner_test.go [hunk 3]
```diff
@@ -576,6 +576,70 @@ func TestParseAndValidateNetCIDR(t *testing.T) {
...
+			cidr:     "192.168.76.0/24",
+			expected: "192.168.76.3",
+		},
+		{
+			name:     "default SLIRP network",
+			cidr:     "10.0.2.0/24",
+			expected: "10.0.2.3",
+		},
+		{
+			name:     "/16 network",
```
Comment (egibs) [inline pkg/container/qemu_runner_test.go:612]: These expected values encode a bug rather than catching one. QEMU assigns `172.16.2.3` for `172.16.0.0/16` and `10.0.2.3` for `10.0.0.0/8`, not `172.16.0.3` / `10.0.0.3`. 

As written the test passes while the real behavior diverges, so it gives false confidence. Worth a case at the /22 boundary (`192.168.0.0/22` → `192.168.2.3`) and the /23 edge case (`192.168.0.0/23` → `192.168.0.3`) as well.

File: pkg/container/qemu_runner.go [hunk 4]
```diff
@@ -2589,6 +2602,40 @@ func parseAndValidateNetCIDR(input string) (string, error) {
...
+		return "", fmt.Errorf("network address is not IPv4")
+	}
+
+	// Convert to uint32 for safe arithmetic, then add 3.
+	n := uint32(base[0])<<24 | uint32(base[1])<<16 | uint32(base[2])<<8 | uint32(base[3])
+	n += 3
+	dns := net.IPv4(byte(n>>24), byte(n>>16), byte(n>>8), byte(n))
+
+	// Verify the result is still within the network.
+	if !ipnet.Contains(dns) {
```
Comment (egibs) [inline pkg/container/qemu_runner.go]: For any prefix ≤ /30 (which `parseAndValidateNetCIDR` enforces), `base+3` is always within the network and for /30 it's the broadcast address, which `Contains` still reports as inside.

File: pkg/container/qemu_runner.go
Comment (egibs) [inline pkg/container/qemu_runner.go]: The /30 lower bound is loose for SLIRP since gateway/DNS/DHCP sit at offsets .2/.3/.15–.31, so a /30 has no room for the DHCP range and `slirpDNSAddr` returns the /30 broadcast (.3) as the DNS. 

Anything narrower than ~/28 won't actually work. It might be worth tightening the minimum (e.g. /29 or /28) so misconfigurations fail fast rather than attempting to use broken networking.

File: pkg/container/qemu_runner.go [hunk 5]
```diff
@@ -2589,6 +2602,40 @@ func parseAndValidateNetCIDR(input string) (string, error) {
...
+// network CIDR. SLIRP uses a fixed offset from the network base: +2 for the
+// gateway and +3 for the DNS server. For example, 192.168.76.0/24 yields
+// 192.168.76.3. The input must be a valid IPv4 CIDR (as returned by
+// parseAndValidateNetCIDR).
+func slirpDNSAddr(cidr string) (string, error) {
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return "", fmt.Errorf("parse CIDR: %w", err)
+	}
+	ipv4 := ip.To4()
```
Comment (egibs) [inline pkg/container/qemu_runner.go]: `ip`/`ipv4` are only used for a nil-check that duplicates the `base := ipnet.IP.To4()` check just below, on input that's already validated as IPv4 upstream so this can probably be dropped.

Also, `binary.BigEndian.Uint32(base)` or `PutUint32` would express the pack/unpack more legibly than the manual shifts.

File: pkg/container/qemu_runner.go (hunk 1 above)
Comment (jmeridth) [reply pkg/container/qemu_runner.go]: @egibs Addressed in https://github.com/chainguard-dev/melange/commit/91c291ac -- instead of trying to match SLIRP internal algorithm, we now set dns= explicitly on the netdev args so melange is authoritative about the DNS address.

File: pkg/container/qemu_runner.go (hunk 2 above)
Comment (jmeridth) [reply pkg/container/qemu_runner.go]: @egibs Addressed in https://github.com/chainguard-dev/melange/commit/91c291ac -- netdevArgs now includes ,dns= alongside ,net= so melange is authoritative and netdev + cmdline always agree.

File: pkg/container/qemu_runner_test.go (hunk 3 above)
Comment (jmeridth) [reply pkg/container/qemu_runner_test.go:612]: @egibs Addressed in https://github.com/chainguard-dev/melange/commit/91c291ac -- added /22, /23, and /28 boundary test cases. Since we now set dns= explicitly on the netdev, the test expectations reflect our chosen address (net+3) rather than QEMU defaults.

File: pkg/container/qemu_runner.go (hunk 4 above)
Comment (jmeridth) [reply pkg/container/qemu_runner.go]: @egibs Addressed in https://github.com/chainguard-dev/melange/commit/91c291ac -- removed the dead Contains check.

File: pkg/container/qemu_runner.go
Comment (jmeridth) [reply pkg/container/qemu_runner.go]: @egibs Addressed in https://github.com/chainguard-dev/melange/commit/91c291ac -- tightened the minimum prefix from /30 to /28. SLIRP needs room for gateway (.2), DNS (.3), and DHCP range (.15+), so anything narrower than /28 cannot work.

File: pkg/container/qemu_runner.go (hunk 5 above)
Comment (jmeridth) [reply pkg/container/qemu_runner.go]: @egibs Addressed in https://github.com/chainguard-dev/melange/commit/91c291ac -- removed the redundant ip/ipv4 check and switched to binary.BigEndian.Uint32/PutUint32.

**Review by egibs — APPROVED** [review]
